syntax = "proto3";

// ----------------------------------------------------------------
// Raft Consensus Protocol Service
// ----------------------------------------------------------------
service RaftService {
  // Leader election RPC
  rpc RequestVote(RequestVoteRequest) returns (RequestVoteResponse);
  
  // Log replication and heartbeat RPC
  rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesResponse);
  
  // Optional: For large state transfers
  rpc InstallSnapshot(InstallSnapshotRequest) returns (InstallSnapshotResponse);
}

// ----------------------------------------------------------------
// RequestVote RPC - Used for leader election
// ----------------------------------------------------------------
message RequestVoteRequest {
  int32 term = 1;                // Candidate's term
  string candidate_id = 2;       // Candidate requesting vote
  int32 last_log_index = 3;      // Index of candidate's last log entry
  int32 last_log_term = 4;       // Term of candidate's last log entry
}

message RequestVoteResponse {
  int32 term = 1;                // Current term, for candidate to update itself
  bool vote_granted = 2;         // True means candidate received vote
}

// ----------------------------------------------------------------
// AppendEntries RPC - Used for log replication and heartbeats
// ----------------------------------------------------------------
message LogEntry {
  int32 term = 1;                // Term when entry was received by leader
  int32 index = 2;               // Position in the log
  string command_type = 3;       // Type of operation (CREATE_USER, POST_MESSAGE, etc.)
  bytes data = 4;                // Serialized command data (JSON)
  string client_id = 5;          // For tracking client requests
  int64 timestamp = 6;           // When command was received
}

message AppendEntriesRequest {
  int32 term = 1;                // Leader's term
  string leader_id = 2;          // So follower can redirect clients
  int32 prev_log_index = 3;      // Index of log entry immediately preceding new ones
  int32 prev_log_term = 4;       // Term of prev_log_index entry
  repeated LogEntry entries = 5; // Log entries to store (empty for heartbeat)
  int32 leader_commit = 6;       // Leader's commit_index
}

message AppendEntriesResponse {
  int32 term = 1;                // Current term, for leader to update itself
  bool success = 2;              // True if follower contained entry matching prev_log_index and prev_log_term
  int32 match_index = 3;         // Highest log entry known to be replicated
  int32 conflict_term = 4;       // Term of conflicting entry (for fast rollback)
  int32 conflict_index = 5;      // First index of conflict_term (for fast rollback)
}

// ----------------------------------------------------------------
// InstallSnapshot RPC - For transferring snapshots (optional)
// ----------------------------------------------------------------
message InstallSnapshotRequest {
  int32 term = 1;                // Leader's term
  string leader_id = 2;          // So follower can redirect clients
  int32 last_included_index = 3; // Snapshot replaces all entries up through this
  int32 last_included_term = 4;  // Term of last_included_index
  bytes data = 5;                // Raw bytes of the snapshot chunk
  bool done = 6;                 // True if this is the last chunk
}

message InstallSnapshotResponse {
  int32 term = 1;                // Current term, for leader to update itself
}

// ----------------------------------------------------------------
// Command Types for State Machine
// ----------------------------------------------------------------
message Command {
  oneof command {
    CreateUserCommand create_user = 1;
    PostMessageCommand post_message = 2;
    CreateGroupCommand create_group = 3;
    AddUserToGroupCommand add_user_to_group = 4;
    RemoveUserFromGroupCommand remove_user_from_group = 5;
    UploadFileMetadataCommand upload_file_metadata = 6;
  }
}

message CreateUserCommand {
  string username = 1;
  string password_hash = 2;
  bool is_admin = 3;
}

message PostMessageCommand {
  string channel_id = 1;
  string sender_id = 2;
  string text = 3;
  string timestamp = 4;
}

message CreateGroupCommand {
  string group_id = 1;
  string group_name = 2;
}

message AddUserToGroupCommand {
  string group_id = 1;
  string user_id = 2;
}

message RemoveUserFromGroupCommand {
  string group_id = 1;
  string user_id = 2;
}

message UploadFileMetadataCommand {
  string channel_id = 1;
  string sender_id = 2;
  string file_name = 3;
  string file_id = 4;
  string timestamp = 5;
}